fsgrid.hpp

```diff
@@ -26,99 +26,193 @@
-                     processDomainDecomposition[0] = i;
-                     processDomainDecomposition[1] = j;
-                     processDomainDecomposition[2] = k;
-                  }
+      }
+      processDomainDecomposition = {1, 1, 1};
+      for (Task_t i = 1; i <= std::min(nProcs, (Task_t)(GlobalSize[0]/minDomainSize[0])); i++) {
+         for (Task_t j = 1; j <= std::min(nProcs, (Task_t)(GlobalSize[1]/minDomainSize[1])) ; j++) {
+            if( i * j  > nProcs ){
+               break;
+            }
+            Task_t k = nProcs/(i*j);
+            // No need to optimize an incompatible DD, also checks for missing remainders
+            if( i * j * k != nProcs ) {
+               continue;
+            }
+
+            processBox[0] = systemDim[0]/i;
+            processBox[1] = systemDim[1]/j;
+            processBox[2] = systemDim[2]/k;
+
+            // Swap the above with this pending a grace period
+            // processBox[0] = calcLocalSize(systemDim[0],i,0);
+            // processBox[1] = calcLocalSize(systemDim[1],j,0);
+            // processBox[2] = calcLocalSize(systemDim[2],k,0);
+
+            int64_t value = 
+               (i > 1 ? processBox[1] * processBox[2]: 0) +
+               (j > 1 ? processBox[0] * processBox[2]: 0) +
+               (k > 1 ? processBox[0] * processBox[1]: 0);
+              
+            // Enable the below pending a grace period
+            // account for singular domains             
+            // if (i!=1 && j!= 1 && k!=1) {
+            //    value *= 13; // 26 neighbours to communicate to
+            // }
+            // if (i==1 && j!= 1 && k!=1) {
+            //    value *= 4; // 8 neighbours to communicate to
+            // }
+            // if (i!=1 && j== 1 && k!=1) {
+            //    value *= 4; // 8 neighbours to communicate to
+            // }
+            // if (i!=1 && j!= 1 && k==1) {
+            //    value *= 4; // 8 neighbours to communicate to
+            // }
+            // else: 2 neighbours to communicate to
+
+            if(value <= optimValue ){
+               optimValue = value;
+               if(value < scored_decompositions.back().first){
+                  scored_decompositions.clear();
                }
+               scored_decompositions.push_back(std::pair<int64_t, std::array<Task_t,3>>(value, {i,j,k}));
             }
          }
+      }
 
-         if(optimValue == std::numeric_limits<double>::max() ||
-               processDomainDecomposition[0] * processDomainDecomposition[1] * processDomainDecomposition[2] != nProcs) {
-            std::cerr << "FSGrid domain decomposition failed, are you running on a prime number of tasks?" << std::endl;
-            throw std::runtime_error("FSGrid computeDomainDecomposition failed");
+      if(myRank == 0 && verbose){
+         std::cout << "(FSGRID) Number of equal minimal-surface decompositions found: " << scored_decompositions.size() << "\n";
+      }
```

Comment:
merge these if-clauses?
(btw I guess MASTER_RANK would be preferable over 0)

Reply:
That's a #define in common.h in Vlasiator, so... hngh, possible define soup.